### PR TITLE
bloom blocks downloading queue

### PR DESCRIPTION
**Pull request by vlad-diachenko**
implemented bloom blocks downloading queue to control the concurrency of downloading the blocks from the storage

### CODE REVIEW

File: pkg/storage/stores/shipper/bloomshipper/block_downloader.go
Comment (chaudum) [inline pkg/storage/stores/shipper/bloomshipper/block_downloader.go:63]: I guess the comment can be removed.

File: pkg/storage/stores/shipper/bloomshipper/block_downloader.go
Comment (chaudum) [inline pkg/storage/stores/shipper/bloomshipper/block_downloader.go:133]: I'm questioning whether we want to spawn a goroutine for each `downloadBlocks` call. Enqueueing is already non-blocking.

I think you can avoid that by returning actual results by the `downloadBlocks()` function (instead of the channels). That would remove the `for { select {} }` in the `ForEachBlock()` function of the shipper.

File: pkg/storage/stores/shipper/bloomshipper/block_downloader.go
Comment (chaudum) [inline pkg/storage/stores/shipper/bloomshipper/block_downloader.go:202]: Maybe add a comment that this function is supposed to close the block.

File: pkg/storage/stores/shipper/bloomshipper/shipper.go
Comment (chaudum) [inline pkg/storage/stores/shipper/bloomshipper/shipper.go:25]: Could be more concise IMO

```suggestion
	BloomShipperDownloadParallelism(tenantID string) int
```

File: pkg/storage/stores/shipper/bloomshipper/block_downloader.go
Comment (owen-d) [reply pkg/storage/stores/shipper/bloomshipper/block_downloader.go:56]: agree, but be careful with the `for ctx.Err` bit as explained in detail [here](https://github.com/grafana/loki/pull/11181#discussion_r1393498266). It should use `select` to avoid saturating the cpu.

**Review by owen-d — COMMENTED** [review]
took a brief look, letting @chaudum do the lions share of review on this one

File: pkg/storage/stores/shipper/bloomshipper/block_downloader.go
Comment (vlad-diachenko) [reply pkg/storage/stores/shipper/bloomshipper/block_downloader.go:10]: done

File: pkg/storage/stores/shipper/bloomshipper/block_downloader.go
Comment (vlad-diachenko) [reply pkg/storage/stores/shipper/bloomshipper/block_downloader.go:8]: done

File: pkg/storage/stores/shipper/bloomshipper/block_downloader.go
Comment (vlad-diachenko) [reply pkg/storage/stores/shipper/bloomshipper/block_downloader.go:21]: done

File: pkg/storage/stores/shipper/bloomshipper/block_downloader.go
Comment (vlad-diachenko) [reply pkg/storage/stores/shipper/bloomshipper/block_downloader.go:59]: I expected queue to be stopped in this case and the worker would receive `ErrStopped` and would close the goroutine 
https://github.com/grafana/loki/blob/27411cff087e6fa2c660c6bfa577c31945ab5232/pkg/queue/queue.go#L146 
wdyt?

Comment (vlad-diachenko) [reply pkg/storage/stores/shipper/bloomshipper/block_downloader.go:59]: ok, but it looks like we do not call StopAsync on the queue ;)

File: pkg/storage/stores/shipper/bloomshipper/block_downloader.go
Comment (vlad-diachenko) [reply pkg/storage/stores/shipper/bloomshipper/block_downloader.go:85]: done

File: pkg/storage/stores/shipper/bloomshipper/block_downloader.go
Comment (vlad-diachenko) [reply pkg/storage/stores/shipper/bloomshipper/block_downloader.go:66]: done

File: pkg/storage/stores/shipper/bloomshipper/block_downloader.go
Comment (vlad-diachenko) [reply pkg/storage/stores/shipper/bloomshipper/block_downloader.go:136]: I want to return the channels because some blocks might be already in the cache, so we can start processing those immediately but for the rest, we will have to download and extract them, which might take some time ... 
wdyt?

File: pkg/storage/stores/shipper/bloomshipper/block_downloader.go
Comment (vlad-diachenko) [reply pkg/storage/stores/shipper/bloomshipper/block_downloader.go]: I managed to achieve the workers shutdown without `select` statement, because `dequeue`  function also checks if the context is not canceled.  
https://github.com/grafana/loki/pull/11201/commits/6b05ec5e236b8747f9c0222bd318f7963419fe7b

File: pkg/storage/stores/shipper/bloomshipper/block_downloader.go
Comment (vlad-diachenko) [reply pkg/storage/stores/shipper/bloomshipper/block_downloader.go:216]: done

File: pkg/storage/stores/shipper/bloomshipper/shipper.go
Comment (vlad-diachenko) [reply pkg/storage/stores/shipper/bloomshipper/shipper.go:25]: yes, it is also correct, but `BloomGatewayBlocksDownloadingParallelism` is more descriptive and explains immediately that it's an option that affects exactly blocks downloading parallelism.  so, I would leave it as is if it's not critical ;)

File: pkg/storage/stores/shipper/bloomshipper/block_downloader.go
Comment (chaudum) [inline pkg/storage/stores/shipper/bloomshipper/block_downloader.go:93]: Could we move this package global callback function into the `blockDownloader` struct?
It would be even easier to use in the tests:

```go
downloader, err := newBlockDownloader(...)
downloader.onWorkerStopCallback = func() { stoppedWorkersCount.Inc() }
```

File: pkg/storage/stores/shipper/bloomshipper/block_downloader.go
Comment (chaudum) [inline pkg/storage/stores/shipper/bloomshipper/block_downloader.go:113]: I would invoke the callback as a deferred function

File: pkg/storage/stores/shipper/bloomshipper/block_downloader.go
Comment (chaudum) [inline pkg/storage/stores/shipper/bloomshipper/block_downloader.go:150]: An exported struct with only non-exported fields is pretty useless for using in another package, unless it implements an interface.
If the type is not used outside of the package, you can make it non-exported.

File: pkg/storage/stores/shipper/bloomshipper/block_downloader.go
Comment (chaudum) [inline pkg/storage/stores/shipper/bloomshipper/block_downloader.go:158]: Not sure if I understand that correctly. On  line 168 you return after sending the first and only error to the `errCh`. Could you just make buffered channel of size `1`?

File: pkg/storage/stores/shipper/bloomshipper/config/config.go
Comment (chaudum) [inline pkg/storage/stores/shipper/bloomshipper/config/config.go:12]: Do we really need the nesting here? Leave that up to you.

File: pkg/storage/stores/shipper/bloomshipper/block_downloader.go
Comment (vlad-diachenko) [reply pkg/storage/stores/shipper/bloomshipper/block_downloader.go:158]: this channel is also passed to the queue and queue workers will report errors to this channel also...

File: pkg/storage/stores/shipper/bloomshipper/block_downloader.go
Comment (vlad-diachenko) [reply pkg/storage/stores/shipper/bloomshipper/block_downloader.go:150]: it is just a leftover, removed it completely

File: pkg/storage/stores/shipper/bloomshipper/block_downloader.go
Comment (vlad-diachenko) [reply pkg/storage/stores/shipper/bloomshipper/block_downloader.go:113]: make sense. done

File: pkg/storage/stores/shipper/bloomshipper/block_downloader.go
Comment (vlad-diachenko) [reply pkg/storage/stores/shipper/bloomshipper/block_downloader.go:93]: make sense ;) thanks

File: pkg/storage/stores/shipper/bloomshipper/block_downloader.go
Comment (vlad-diachenko) [reply pkg/storage/stores/shipper/bloomshipper/block_downloader.go:95]: done

File: pkg/storage/stores/shipper/bloomshipper/config/config.go
Comment (vlad-diachenko) [reply pkg/storage/stores/shipper/bloomshipper/config/config.go:12]: I believe that it will be better to have it separated ;) also do not have a strong opinion on it... just separated it because in the next PR I will add cache config and do not want to mix the properties in the single block....

File: pkg/storage/stores/shipper/bloomshipper/block_downloader.go
Comment (vlad-diachenko) [reply pkg/storage/stores/shipper/bloomshipper/block_downloader.go]: fixed

File: pkg/storage/stores/shipper/bloomshipper/block_downloader.go
Comment (vlad-diachenko) [reply pkg/storage/stores/shipper/bloomshipper/block_downloader.go:108]: it's not necessary here because the context will be checked during `Dequeue` function call...
also, we have discussed it here https://raintank-corp.slack.com/archives/C056KUQFBFT/p1700665187391789

File: pkg/storage/stores/shipper/bloomshipper/client.go
Comment (vlad-diachenko) [reply pkg/storage/stores/shipper/bloomshipper/client.go:194]: we will download many blocks in parallel using the downloading queue. I see that in the compactor it might be necessary to download many blocks in parallel also, but I believe we can spin up the goroutines in compactor's code to download many blocks in parallel. wdyt?

File: pkg/storage/stores/shipper/bloomshipper/client.go
Comment (vlad-diachenko) [reply pkg/storage/stores/shipper/bloomshipper/client.go:200]: it's already handled in https://github.com/grafana/loki/blob/d96c9396933ebf015568c7d4a7ff0f4568915273/pkg/storage/stores/shipper/bloomshipper/block_downloader.go#L134

File: pkg/bloomgateway/bloomgateway_test.go
Comment (chaudum) [inline pkg/bloomgateway/bloomgateway_test.go:257]:
```suggestion
	return 0
```

File: pkg/bloomgateway/bloomgateway_test.go
Comment (chaudum) [inline pkg/bloomgateway/bloomgateway_test.go:262]:
```suggestion
	return true
```

File: pkg/bloomgateway/bloomgateway_test.go
Comment (chaudum) [inline pkg/bloomgateway/bloomgateway_test.go:267]:
```suggestion
	return 1
```

File: pkg/storage/stores/shipper/bloomshipper/block_downloader.go
Comment (chaudum) [inline pkg/storage/stores/shipper/bloomshipper/block_downloader.go:205]: Maybe we should log the error?

File: pkg/storage/stores/shipper/bloomshipper/config/config.go
Comment (chaudum) [inline pkg/storage/stores/shipper/bloomshipper/config/config.go:21]: I think we should be more conservative regarding the default value. `10` should  be more than sufficient?

File: pkg/storage/stores/shipper/bloomshipper/config/config.go
Comment (chaudum) [inline pkg/storage/stores/shipper/bloomshipper/config/config.go:22]: Same here.

**Review by chaudum — APPROVED** [review]
Left a few comments, but nothing that would block merging.

**Pull request merged**
